core/shared/src/main/scala/fs2/Hotswap.scala

```diff
@@ -0,0 +1,118 @@
+import cats.effect.concurrent.Ref
+import cats.effect.implicits._
+
+/**
+  * Supports treating a linear sequence of resources as a single resource.
+  *
+  * A `Hotswap[F, R]` instance is created as a `Resource` and hence, has
+  * a lifetime that is scoped by the `Resource`. After creation, a `Resource[F, R]`
+  * can be swapped in to the `Hotswap` by calling `swap`. The acquired resource
+  * is returned and is finalized when the `Hotswap` is finalized or upon the next
+  * call to `swap`, whichever occurs first.
+  *
+  * For example, the sequence of three resources `r1, r2, r3` are shown in the
+  * following diagram:
+  *
+  * {{{
+  * >----- swap(r1) ---- swap(r2) ---- swap(r3) ----X
+  * |        |             |             |          |
+  * Creation |             |             |          |
+  *         r1 acquired    |             |          |
+  *                       r2 acquired    |          |
+  *                       r1 released   r3 acquired |
+  *                                     r2 released |
+  *                                                r3 released
+  * }}}
+  *
+  * This class is particularly useful when working with pulls that cycle through
+  * resources -- e.g., writing bytes to files, rotating files every N bytes or M seconds.
+  * Without `Hotswap`, such pulls leak resources -- on each file rotation, a file handle
+  * or at least an internal resource reference accumulates. With `Hotswap`, the `Hotswap`
+  * instance is the only registered resource and each file is swapped in to the `Hotswap`.
+  *
+  * Usage typically looks something like:
+  *
+  * {{{
+  * Stream.resource(Hotswap(mkResource)).flatMap { case (hotswap, r) =>
+  *   // Use r, call hotswap.swap(mkResource) as necessary
+  * }
+  * }}}
+  *
+  * See `fs2.io.file.writeRotate` for an example of usage.
+  */
+sealed trait Hotswap[F[_], R] {
+
+  /**
+    * Allocates a new resource, closes the last one if present, and
+    * returns the newly allocated `R`.
+    *
+    * If there are no further calls to `swap`, the resource created by
+    * the last call will be finalized when the lifetime of
+    * this `Hotswap` (which is itself tracked by `Resource`) is over.
+    *
+    * Since `swap` closes the old resource immediately, you need to
+    * ensure that no code is using the old `R` when `swap` is called.
+    * Failing to do so is likely to result in an error on the
+    * _consumer_ side. In any case, no resources will be leaked by
+    * `swap`.
+    *
+    * If you try to call swap after the lifetime of this `Hotswap` is
+    * over, `swap` will fail, but it will ensure all resources are
+    * closed, and never leak any.
+    */
+  def swap(next: Resource[F, R]): F[R]
+}
+
+object Hotswap {
+
+  /**
+    * Creates a new `Hotswap` initialized with the specified resource.
+    * The `Hotswap` instance and the initial resource are returned.
+    */
+  def apply[F[_]: Concurrent, R](initial: Resource[F, R]): Resource[F, (Hotswap[F, R], R)] =
```

Comment:
how inconvenient do you think it would be to rename this to `create` or `make`, and `create` to `empty`? `apply` is pretty much as convenient as the Stream combinator you wanted, so if you really want that convenience, ignore this comment. My doubt stems from the fact that `apply` on tagless stuff is generally for summoning

Reply:
Good point, leave it